router/fastdp.go

```diff
@@ -0,0 +1,1115 @@
+
+	// the only special packet type is a heartbeat
+	if len(frame) < EthernetOverhead+10 {
+		log.Warning(fwd.logPrefix(), "short vxlan special packet: ", len(frame), " bytes")
+		return
+	}
+
+	if binary.BigEndian.Uint64(frame[EthernetOverhead:]) != fwd.connUID ||
+		uint16(len(frame)) != binary.BigEndian.Uint16(frame[EthernetOverhead+8:]) {
+		return
+	}
+
+	if fwd.remoteAddr == nil {
+		fwd.remoteAddr = sender
+
+		if fwd.confirmed {
+			fwd.heartbeatTimer.Reset(0)
+		}
+	} else if !udpAddrsEqual(fwd.remoteAddr, sender) {
+		log.Info(fwd.logPrefix(),
+			"Peer IP address changed to ", sender)
+		fwd.remoteAddr = sender
+	}
+
+	if !fwd.ackedHeartbeat {
+		fwd.ackedHeartbeat = true
+		fwd.handleError(fwd.sendControlMsg([]byte{HeartbeatAck}))
+	}
+
+	// we can receive a heartbeat before Confirm() has set up
+	// heartbeatTimeout
+	if fwd.heartbeatTimeout != nil {
+		fwd.heartbeatTimeout.Reset(HeartbeatTimeout)
+	}
+}
+
+func (fwd *fastDatapathForwarder) ControlMessage(msg []byte) {
+	fwd.lock.Lock()
+	defer fwd.lock.Unlock()
+
+	if len(msg) == 0 {
+		log.Info(fwd.logPrefix(),
+			"Received zero-length control message")
+	}
+
+	switch msg[0] {
+	case HeartbeatAck:
+		fwd.handleHeartbeatAck()
+
+	default:
+		log.Info(fwd.logPrefix(),
+			"Ignoring unknown control message: ", msg[0])
+	}
+}
+
+func (fwd *fastDatapathForwarder) handleHeartbeatAck() {
+	log.Debug(fwd.logPrefix(), "handleHeartbeatAck")
+
+	if fwd.heartbeatInterval != SlowHeartbeat {
+		close(fwd.establishedChan)
+		fwd.heartbeatInterval = SlowHeartbeat
+		if fwd.heartbeatTimer != nil {
+			fwd.heartbeatTimer.Reset(fwd.heartbeatInterval)
+		}
+	}
+}
+
+func (fwd *fastDatapathForwarder) Forward(key ForwardPacketKey) FlowOp {
+	fwd.lock.RLock()
+	defer fwd.lock.RUnlock()
+
+	if fwd.remoteAddr == nil {
+		// Returning nil would discard the packet, but also
+		// result in a flow rule, which we would have to
+		// invalidate when we learn the remote IP.  So for
+		// now, just prevent flows.
+		return vetoFlowCreationFlowOp{}
+	}
+
+	remoteIP, err := ipv4Bytes(fwd.remoteAddr.IP)
+	if err != nil {
+		log.Error(err)
+		return nil
+	}
+
+	var sta odp.SetTunnelAction
+	sta.SetTunnelId(tunnelIDFor(key))
+	sta.SetIpv4Src(fwd.localIP)
+	sta.SetIpv4Dst(remoteIP)
+	sta.SetTos(0)
+	sta.SetTtl(64)
+	sta.SetDf(true)
+	sta.SetCsum(false)
+	return fwd.fastdp.odpActions(sta,
+		odp.NewOutputAction(fwd.fastdp.mainVxlanVportID))
+}
+
+func tunnelIDFor(key ForwardPacketKey) (tunnelID [8]byte) {
+	src := uint64(key.SrcPeer.ShortID)
+	dst := uint64(key.DstPeer.ShortID)
+	binary.BigEndian.PutUint64(tunnelID[:], src|dst<<12)
+	return
+}
+
+func (fwd *fastDatapathForwarder) Stop() {
+	// Might be nice to delete all the relevant flows here, but we
+	// can just let them expire.
+	fwd.fastdp.removeForwarder(fwd.remotePeer.Name, fwd)
+
+	fwd.lock.Lock()
+	defer fwd.lock.Unlock()
+	fwd.sendControlMsg = func([]byte) error { return nil }
+
+	select {
+	case fwd.errorChan <- nil:
```

Comment:
I can't fathom the purpose of this - we don't do it in `sleeveForwarder.Stop()` either